---
layout: article
title: Spring 事务传播行为
tags: ["Spring", "事务"]
---

## PROPAGATION_REQUIRED（默认）

> 支持当前事务，如果不存在则创建一个新事务

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional()
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
@Transactional
public void test() {
    serviceA.method();
    serviceB.exception_method();
}
```

- 输出日志

```
2021-03-22 16:51:45.705 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating new transaction with name [com.example.testspringtrancationpropagation.test.PropagationRequiredService.test]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2021-03-22 16:51:45.705 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@1583722422 wrapping com.mysql.cj.jdbc.ConnectionImpl@5b3a7ef5] for JDBC transaction
2021-03-22 16:51:45.708 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@1583722422 wrapping com.mysql.cj.jdbc.ConnectionImpl@5b3a7ef5] to manual commit
2021-03-22 16:51:45.722 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating in existing transaction
2021-03-22 16:51:45.754 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating in existing transaction
2021-03-22 16:51:45.763 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating transaction failed - marking existing transaction as rollback-only
2021-03-22 16:51:45.763 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Setting JDBC transaction [HikariProxyConnection@1583722422 wrapping com.mysql.cj.jdbc.ConnectionImpl@5b3a7ef5] rollback-only
2021-03-22 16:51:45.763 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction rollback
2021-03-22 16:51:45.763 DEBUG 4070 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back JDBC transaction on Connection [HikariProxyConnection@1583722422 wrapping com.mysql.cj.jdbc.ConnectionImpl@5b3a7ef5]
```

*创建 PropagationRequiredService.test 事务，serviceA.method()、serviceB.exception_method() 加入到当前事务里、serviceB.exception_method() 抛出异常，事务回滚。*

### 测试 2

- 代码

```java
@Transactional
public void test3() {
    serviceA.method();
    try {
        serviceB.exception_method();
    } catch (Exception e) {
        log.info("捕获异常 ==> {}", e.getMessage());
    }
}
```

- 日志

```
021-03-22 17:03:15.704 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating new transaction with name [com.example.testspringtrancationpropagation.test.PropagationRequiredService.test3]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2021-03-22 17:03:15.705 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@886386381 wrapping com.mysql.cj.jdbc.ConnectionImpl@5d21202d] for JDBC transaction
2021-03-22 17:03:15.706 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@886386381 wrapping com.mysql.cj.jdbc.ConnectionImpl@5d21202d] to manual commit
2021-03-22 17:03:15.720 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating in existing transaction
2021-03-22 17:03:15.743 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating in existing transaction
2021-03-22 17:03:15.751 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Participating transaction failed - marking existing transaction as rollback-only
2021-03-22 17:03:15.751 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Setting JDBC transaction [HikariProxyConnection@886386381 wrapping com.mysql.cj.jdbc.ConnectionImpl@5d21202d] rollback-only
2021-03-22 17:03:15.751  INFO 4102 --- [           main] c.e.t.test.PropagationRequiredService    : 捕获异常 ==> null
2021-03-22 17:03:15.752 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Global transaction is marked as rollback-only but transactional code requested commit
2021-03-22 17:03:15.752 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction rollback
2021-03-22 17:03:15.752 DEBUG 4102 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back JDBC transaction on Connection [HikariProxyConnection@886386381 wrapping com.mysql.cj.jdbc.ConnectionImpl@5d21202d]
```

*虽然抛出的异常被 catch 了，但是在当前事务还是标记为回滚。*

## PROPAGATION_SUPPORTS

> 支持当前事务，如果不存在则非事务执行

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.SUPPORTS)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.SUPPORTS)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.SUPPORTS)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.SUPPORTS)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
public void test() {
    serviceA.method();
    serviceB.exception_method();
}
```

- 输出日志

```
2021-03-23 09:59:07.653 DEBUG 4976 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Should roll back transaction but cannot - no transaction available
```

*serviceA.method(); 成功提交，serviceB.exception_method(); 抛出异常也成功提交，当前没有事务则使用非事务方法运行*。如果当前方法已有事务则和 PROPAGATION_REQUIRED 测试 1、测试 2 的结果一样。

## PROPAGATION_MANDATORY

> 支持当前事务，如果不存在则抛出异常

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.MANDATORY)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.MANDATORY)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.MANDATORY)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.MANDATORY)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
public void test() {
    serviceA.method();
    serviceB.method();
}
```

- 输出日志

```
org.springframework.transaction.IllegalTransactionStateException: No existing transaction found for transaction marked with propagation 'mandatory'
```

*抛出 IllegalTransactionStateException 异常，此传播级别的事务只能在父事务中运行，否则抛出异常。如果当前方法已有事务则和 PROPAGATION_REQUIRED 测试 1、测试 2 的结果一样。*

## PROPAGATION_REQUIRES_NEW

> Create a new transaction, and suspend the current transaction if one exists.
>
> 创建一个新事务，并暂停当前事务（如果存在）。

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.REQUIRES_NEW)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
@Transactional
public void test2() {
    jdbcTemplate.update("insert into test(name) value ('PropagationRequiresNewService')");
    serviceA.method();
    serviceB.exception_method();
}
```

- 输出日志

```
2021-03-23 10:44:02.826 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating new transaction with name [com.example.testspringtrancationpropagation.test.PropagationRequiresNewService.test2]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2021-03-23 10:44:02.826 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@576222471 wrapping com.mysql.cj.jdbc.ConnectionImpl@5c77ba8f] for JDBC transaction
2021-03-23 10:44:02.828 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@576222471 wrapping com.mysql.cj.jdbc.ConnectionImpl@5c77ba8f] to manual commit
2021-03-23 10:44:02.875 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Suspending current transaction, creating new transaction with name [com.example.testspringtrancationpropagation.service.ServiceA.method]
2021-03-23 10:44:02.876 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@71209097 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] for JDBC transaction
2021-03-23 10:44:02.876 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@71209097 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] to manual commit
2021-03-23 10:44:02.884 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction commit
2021-03-23 10:44:02.884 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Committing JDBC transaction on Connection [HikariProxyConnection@71209097 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8]
2021-03-23 10:44:02.886 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Releasing JDBC Connection [HikariProxyConnection@71209097 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] after transaction
2021-03-23 10:44:02.886 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Resuming suspended transaction after completion of inner transaction
2021-03-23 10:44:02.886 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Suspending current transaction, creating new transaction with name [com.example.testspringtrancationpropagation.service.ServiceB.exception_method]
2021-03-23 10:44:02.886 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@2058127492 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] for JDBC transaction
2021-03-23 10:44:02.887 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@2058127492 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] to manual commit
2021-03-23 10:44:02.893 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction rollback
2021-03-23 10:44:02.893 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back JDBC transaction on Connection [HikariProxyConnection@2058127492 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8]
2021-03-23 10:44:02.894 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Releasing JDBC Connection [HikariProxyConnection@2058127492 wrapping com.mysql.cj.jdbc.ConnectionImpl@3c5dbdf8] after transaction
2021-03-23 10:44:02.894 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Resuming suspended transaction after completion of inner transaction
2021-03-23 10:44:02.894 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction rollback
2021-03-23 10:44:02.895 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back JDBC transaction on Connection [HikariProxyConnection@576222471 wrapping com.mysql.cj.jdbc.ConnectionImpl@5c77ba8f]
2021-03-23 10:44:02.896 DEBUG 5266 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Releasing JDBC Connection [HikariProxyConnection@576222471 wrapping com.mysql.cj.jdbc.ConnectionImpl@5c77ba8f] after transaction

```

*serviceA.method(); 提交成功，jdbcTemplate.update()、serviceB.exception_method(); 回滚，内部的事务是相互独立的，但是内部事务抛出异常会影响当前事务，当前事务则不会影响内部事务。*

## PROPAGATION_NOT_SUPPORTED

> Execute non-transactionally, suspend the current transaction if one exists.
>
> 非事务执行，如果存在当前事务，则挂起当前事务。

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NOT_SUPPORTED)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.NOT_SUPPORTED)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NOT_SUPPORTED)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.NOT_SUPPORTED)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
@Transactional
public void test() {
    jdbcTemplate.update("insert into test(name) value ('PROPAGATION_NOT_SUPPORTED')");
    serviceA.method();
    serviceB.exception_method();
}
```

- 输出日志

```
2021-03-23 11:03:09.380 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating new transaction with name [com.example.testspringtrancationpropagation.test.PropagationNotSupportsService.test2]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2021-03-23 11:03:09.381 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@1338229093 wrapping com.mysql.cj.jdbc.ConnectionImpl@10a0fe30] for JDBC transaction
2021-03-23 11:03:09.385 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@1338229093 wrapping com.mysql.cj.jdbc.ConnectionImpl@10a0fe30] to manual commit
2021-03-23 11:03:09.480 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Suspending current transaction
2021-03-23 11:03:09.500 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Resuming suspended transaction after completion of inner transaction
2021-03-23 11:03:09.500 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Suspending current transaction
2021-03-23 11:03:09.593 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Should roll back transaction but cannot - no transaction available
2021-03-23 11:03:09.594 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Resuming suspended transaction after completion of inner transaction
2021-03-23 11:03:09.594 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction rollback
2021-03-23 11:03:09.594 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back JDBC transaction on Connection [HikariProxyConnection@1338229093 wrapping com.mysql.cj.jdbc.ConnectionImpl@10a0fe30]
2021-03-23 11:03:09.595 DEBUG 5333 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Releasing JDBC Connection [HikariProxyConnection@1338229093 wrapping com.mysql.cj.jdbc.ConnectionImpl@10a0fe30] after transaction
```

*serviceA.method();  serviceB.exception_method(); 提交成功，jdbcTemplate.update() 回滚。内部方法以非事务执行，当前方法的事务会挂起，正常的提交或抛出异常回滚。*

## PROPAGATION_NEVER

> Execute non-transactionally, throw an exception if a transaction exists.
>
> 非事务执行，如果存在事务，则引发异常。

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NEVER)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.NEVER)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NEVER)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.NEVER)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
public void test() {
    serviceA.method();
    serviceB.exception_method();
}
```

- 输出日志

```
2021-03-23 11:14:09.733 DEBUG 5365 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Should roll back transaction but cannot - no transaction available
```

*serviceA.method();  serviceB.exception_method(); 提交成功，内部方法以非事务方法执行。如当前方法加上事务则抛出异常*

## PROPAGATION_NESTED

> Execute within a nested transaction if a current transaction exists, behave like REQUIRED otherwise.
>
> 如果当前事务存在，则在嵌套事务中执行，否则，行为类似于REQUIRED。

- ServiceA

```java
@Service
public class ServiceA {

    @Resource
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NESTED)
    public void method() {
        jdbcTemplate.update("insert into `test`(name) values ('zhangsan')");
    }

    @Transactional(propagation = Propagation.NESTED)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) values ('zhangsan2')");
        throw new RuntimeException();
    }
}
```

- ServiceB

```java
@Service
public class ServiceB {

    @Autowired
    private JdbcTemplate jdbcTemplate;

    @Transactional(propagation = Propagation.NESTED)
    public void method() {
        jdbcTemplate.update("insert into test(name) value ('lisi1')");
    }

    @Transactional(propagation = Propagation.NESTED)
    public void exception_method() {
        jdbcTemplate.update("insert into test(name) value ('lisi2')");
        throw new RuntimeException();
    }
}
```

### 测试 1

- 代码

```java
@Transactional
public void test3() {
    serviceA.method();
    try {
        serviceB.exception_method();
    } catch (Exception e) {
        log.info("捕获异常 ==> {}", e.getMessage());
    }
}
```

- 输出日志

```
2021-03-23 11:37:36.493 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating new transaction with name [com.example.testspringtrancationpropagation.test.PropagationNestedService.test3]: PROPAGATION_REQUIRED,ISOLATION_DEFAULT
2021-03-23 11:37:36.494 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Acquired Connection [HikariProxyConnection@1987707214 wrapping com.mysql.cj.jdbc.ConnectionImpl@d109c4f] for JDBC transaction
2021-03-23 11:37:36.495 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Switching JDBC Connection [HikariProxyConnection@1987707214 wrapping com.mysql.cj.jdbc.ConnectionImpl@d109c4f] to manual commit
2021-03-23 11:37:36.510 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating nested transaction with name [com.example.testspringtrancationpropagation.service.ServiceA.method]
2021-03-23 11:37:36.597 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Releasing transaction savepoint
2021-03-23 11:37:36.597 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Creating nested transaction with name [com.example.testspringtrancationpropagation.service.ServiceB.exception_method]
2021-03-23 11:37:36.603 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Rolling back transaction to savepoint
2021-03-23 11:37:36.604  INFO 5531 --- [           main] c.e.t.test.PropagationNestedService      : 捕获异常 ==> null
2021-03-23 11:37:36.604 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Initiating transaction commit
2021-03-23 11:37:36.604 DEBUG 5531 --- [           main] o.s.jdbc.support.JdbcTransactionManager  : Committing JDBC transaction on Connection [HikariProxyConnection@1987707214 wrapping com.mysql.cj.jdbc.ConnectionImpl@d109c4f]
```

*serviceA.method(); 提交成功，serviceB.exception_method(); 回滚。内部事务嵌套在当前事务中执行，如当前事务回滚，内部事务也回滚，反之内部事务抛出异常被 catch 没有被外部事务感知则不会回滚*

